extendr-api/src/ownership.rs

```diff
@@ -58,22 +100,63 @@ impl Ownership {
             let preservation = Rf_allocVector(VECSXP, INITIAL_PRESERVATION_SIZE as R_xlen_t);
             R_PreserveObject(preservation);
             Ownership {
-                preservation: preservation as usize,
+                preservation: preservation.into(),
                 cur_index: 0,
                 max_index: INITIAL_PRESERVATION_SIZE,
                 objects: HashMap::with_capacity(INITIAL_PRESERVATION_SIZE),
             }
         }
     }
 
+    // Garbage collect the tracking structures.
+    unsafe fn garbage_collect(&mut self) {
+        // println!("garbage_collect {} {}", self.cur_index, self.max_index);
+        let new_size = self.cur_index * 2 + EXTRA_PRESERVATION_SIZE;
+        let new_sexp = Rf_allocVector(VECSXP, new_size as R_xlen_t);
+        R_PreserveObject(new_sexp);
+        let old_sexp = self.preservation.inner();
+
+        let mut new_objects = HashMap::with_capacity(new_size);
+
+        // copy non-null elements to new vector and hashmap.
+        let mut j = 0;
+        for (addr, object) in self.objects.iter() {
+            if object.refcount != 0 {
+                SET_VECTOR_ELT(new_sexp, j as R_xlen_t, addr.inner());
+                new_objects.insert(
+                    addr.clone(),
+                    Object {
+                        refcount: object.refcount,
+                        index: j,
+                    },
+                );
+                j += 1;
+            }
+        }
+
+        R_ReleaseObject(old_sexp);
+        self.preservation.set_inner(new_sexp);
```

Comment:
I would suggest to instead create new `SendSEXP` and assign it to `.preservation`. This will allow discarding `set_inner()` function in favor of some constructor.